server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -193,6 +201,23 @@ public void allocate(RoutingAllocation allocation, ActionListener<Void> listener
         reconcile(currentDesiredBalance, allocation);
     }
 
+    private synchronized void processNodeShutdowns(ClusterState clusterState) {
+        // Clean up processed shutdowns that are removed from the cluster metadata
+        processedNodeShutdowns.removeIf(nodeId -> clusterState.metadata().nodeShutdowns().contains(nodeId) == false);
```

Comment:
since we only run this on reroute and not on all cluster state updates, the removal might happen at some point later, since a reroute doesn't happen when we delete the shutdown marker.

Reply:
Although, if the next time this is called is when we have marked the node for shutdown again after removing the marker, we wouldn't reset again! Not sure, is that an issue?

Reply:
there seems to be a `startedAtMillis` in each`SingleNodeShutdownMetadata`. we could use that + the node id.

Reply:
I think it's not an issue, if we never triggered a computation with the shutdown marker removed then there has been no substantial change to the shard allocation, no need to reset.

Reply:
Thanks. Then, I guess this is ready for review.